drivers/regulator/regulator_pca9422.c

```diff
@@ -0,0 +1,1234 @@
+	}
+
+	/* have to unlock first or else register read returns 0x0 */
+	ret = regulator_pca9422_unlock(&cconfig->i2c);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (enmode_val) {
+		/* Configuration has this regulator disabled in some modes. */
+		if (enmode_val == PCA9422_ENMODE_ENABLED_NONE) {
+			/* Disabling in all modes can only be done via enable_reg */
+			ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+						     config->desc->enable_reg,
+						     config->desc->enable_mask, 0U);
+		} else {
+			/*
+			 * Disabling in non ACTIVE modes can be done via
+			 * the mode_reg.
+			 */
+			ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+						     config->desc->mode_reg,
+						     config->desc->enmode_mask,
+						     enmode_val << config->desc->enmode_pos);
+		}
+
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	/* configure mode voltages */
+	for (size_t i = 0U; i < ARRAY_SIZE(config->modes_uv); i++) {
+		if (config->modes_uv[i] > 0) {
+			uint16_t idx;
+
+			/* program mode voltage */
+			ret = linear_range_group_get_win_index(
+				config->desc->ranges, config->desc->num_ranges,
+				config->modes_uv[i], config->modes_uv[i], &idx);
+			if (ret == -EINVAL) {
+				LOG_ERR("invalid voltage %u", config->modes_uv[i]);
+				goto out;
+			}
+
+			ret = i2c_reg_update_byte_dt(
+				&cconfig->i2c,
+				config->desc->vsel_mode[i].vsel_reg,
+				config->desc->vsel_mode[i].vsel_mask, (uint8_t)idx);
+			if (ret < 0) {
+				goto out;
+			}
+		}
+	}
+
+	/* configure lpmode */
+	if (config->desc->lpmode_mask) {
+		ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+					     config->desc->mode_reg,
+					     config->desc->lpmode_mask,
+					     config->lpmode << config->desc->lpmode_pos);
+		if (ret < 0) {
+			goto out;
+		}
+	} else {
+		/* LDO1 and LDO4 have no LPMODE control register. Value should be 3. */
+		assert(config->lpmode == 3);
+	}
+
+out:
+	regulator_pca9422_lock(&cconfig->i2c);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return regulator_common_init(dev, false);
+}
+
+int regulator_pca9422_dvs_state_set(const struct device *dev,
+				    regulator_dvs_state_t state)
+{
+	struct regulator_pca9422_common_data *data = dev->data;
+
+	if (state >= PCA9422_NUM_MODES) {
+		return -ENOTSUP;
+	}
+
+	/*
+	 * The user cannot set DVS state via this API,
+	 * but they may want to query/set voltages for another mode.
+	 * Return -EPERM to indicate change failed, but change the
+	 * dvs_state variable so the user can access the alternative
+	 * dvs mode settings.
+	 */
+	data->dvs_state = state;
+	return -EPERM;
+}
+
+static const struct regulator_parent_driver_api parent_api = {
+	.dvs_state_set = regulator_pca9422_dvs_state_set,
+};
+
+static int regulator_pca9422_common_init(const struct device *dev)
+{
+	const struct regulator_pca9422_common_config *config = dev->config;
+	uint8_t reg_val;
+	int ret;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		return -ENODEV;
+	}
+
+	/* configure VIN current limit */
+	if (config->vin_ilim_ua < PCA9422_CHGIN_IN_LIMIT_MIN_BIG_UA) {
```

Comment:
As far as I can see this is set by an enum, so you may be able to use DT_INST_ENUM_IDX to calculate the required register value at compile time.

Reply:
Done. Thanks for the suggestion.